core/src/main/java/org/jruby/RubyThread.java

```diff
@@ -382,8 +377,16 @@ public IRubyObject setErrorInfo(IRubyObject errorInfo) {
         return errorInfo;
     }
 
-    public void setContext(ThreadContext context) {
-        this.contextRef = new WeakReference<ThreadContext>(context);
+    public void setContext(final ThreadContext context) {
+        if (context == null) {
+            WeakReference<ThreadContext> contextRef = this.contextRef;
+            if (contextRef != null) {
+                contextRef.clear();
+                this.contextRef = null;
```

Comment:
Maybe this should be moved to a "clearContext" method rather than special-casing null?

Reply:
good point, let's have a `clearContext`